### PR TITLE
Vector Similarity data type

### PR DESCRIPTION
**Pull request by ashtul**
**Creation**:
`FT.CREATE idx SCHEMA vecsim VECTOR type size metric algorithm INITIAL_CAP {cap} M {m} EF {ef}`
* type: FLOAT32, FLOAT64, INT32, INT64
* size: size of vector
* metric: IP, L2
* algorithm: BF, HNSW

Example: `FT.CREATE idx SCHEMA vecsim VECTOR FLOAT32 16 L2 HNSW INITIAL_CAP 10 M 16 EF 200`

**Query**:
`FT.SEARCH idx @vecsim:[value query count]`
* value: the vector
* query: TOPK, RANGE
* count: k for TOPK, distance for RANGE

Example: `FT.SEARCH idx @vecsim:[abcdefgh TOPK 4]`

### CODE REVIEW

File: src/vector_index.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+typedef enum {
+  VECTOR_TOPK = 0,
+  VECTOR_RANGE = 1,
+} VectorQueryType;
+
+typedef enum {
+  VECTOR_DIST_L2 = 0,
+  VECTOR_DIST_COSINE = 1,
+  VECTOR_DIST_IP = 2,
+} VectorDistanceMetric;
```
Comment (rafie) [inline src/vector_index.h:44]: These guys are defined in `VecSim/vecsim.h`. Do we need to re-define them here?

File: src/vector_index.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  VecSimIndex *data;
+
+  size_t size;
+  VectorType type;
+  VectorAlgorithm alg;
+  VectorDistanceMetric dist;
+
+  size_t numEntries;
+  t_docId lastDocId;
+
```
Comment (rafie) [inline src/vector_index.h:74]: Remove?

File: src/document.c [hunk 3]
```diff
@@ -471,6 +472,35 @@ FIELD_BULK_INDEXER(numericIndexer) {
...
 
+FIELD_PREPROCESSOR(vectorPreprocessor) {
+  // TODO: check input validity
+  size_t len;
+  fdata->vector = RedisModule_StringPtrLen(field->text, &len);
+  fdata->vecLen = len;
+  return 0;
+}
+
+int HNSWIndex_AddVector(VecSimIndex *index, const void* vector_data, size_t id);
```
Comment (rafie) [inline src/document.c:483]: Is this a thing?

Comment (ashtul) [reply src/document.c:483]: removed

File: src/vector_index.h (hunk 1 above)
Comment (ashtul) [reply src/vector_index.h:44]: removed

File: src/index.c [hunk 4]
```diff
@@ -1962,6 +1963,7 @@ void Profile_AddIters(IndexIterator **root) {
       }
       break;
     case READ_ITERATOR:
+    case LIST_ITERATOR:
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/index.c:1966]: Leave as-is for now

File: src/vector_index.c [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+      // Note: RANGE is being used for base64 conversion queries.
+      vector = base64_decode(vf->vector, vf->vecLen, &outLen);
+      // TODO: check outLen == expected len.
+      vf->results = VecSimIndex_TopKQuery(vecsim, vector, vf->value);
+      rm_free(vector);
+      break;
+  }
+
+  return NewListIterator(vf->results, vf->value);
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/vector_index.c:65]: Awaiting support on VecSim API.

**Pull request merged**